contracts/src/types/V2.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.25;
+
+import {ParaID} from "./Common.sol";
+
+struct Origin {
+    // Origin network
+    uint8 network;
+    // Origin chain
+    ParaID paraID;
+    // Stable hash of full origin location;
+    bytes32 locationID;
+}
+
+// Inbound message from a Polkadot parachain (via BridgeHub)
+struct InboundMessage {
+    // origin
+    bytes32 origin;
+    // Message nonce
+    uint64 nonce;
+    // Commands
+    Command[] commands;
+}
```

Comment:
Do we need `messge_id` or `tx_hash`  generated on source chain for tracing purpose?

Reply:
I think `nonce` is sufficient for tracing purposes, when one takes this flow into account.

1. BH receives XCM
2. BH processes XCM and emits `MessageProcessed(xcm_id, nonce)`
3. Gateway receives message and logs the nonce.

Then the UX layer and other indexers can use the nonce logged in 2 and 3 to link the message to the XCM topic or hash.